data_model/src/transaction.rs

```diff
@@ -342,34 +350,53 @@ impl Transaction {
         self.payload.nonce = Some(nonce);
         self
     }
+}
 
-    /// Sign transaction with the provided key pair.
-    ///
-    /// # Errors
-    ///
-    /// Fails if signature creation fails
-    #[cfg(feature = "std")]
-    pub fn sign(
-        mut self,
-        key_pair: iroha_crypto::KeyPair,
-    ) -> Result<Transaction, iroha_crypto::Error> {
-        let signature = SignatureOf::new(key_pair, &self.payload)?;
-        self.signatures.insert(signature);
-
-        Ok(Self {
-            payload: self.payload,
-            signatures: self.signatures,
-        })
-    }
+/// Structure represents transaction in non-trusted form.
+///
+/// `Iroha` and its' clients use [`SignedTransaction`] to send transactions
+/// via network.  Direct usage in business logic is strongly
+/// prohibited. Before any interactions `accept`.
+#[version(n = 1, versioned = "VersionedSignedTransaction")]
+#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+pub struct SignedTransaction {
+    /// [`Transaction`] payload.
+    pub payload: Payload,
+    /// [`SignatureOf`] [`Payload`].
+    pub signatures: btree_set::BTreeSet<SignatureOf<Payload>>,
 }
 
-impl Txn for Transaction {
-    type HashOf = Self;
+macro_rules! impl_tx {
+    ($($ty:ty),+$(,)?) => {$(
+        impl Txn for $ty {
+            type HashOf = SignedTransaction;
 
-    #[inline]
-    fn payload(&self) -> &Payload {
-        &self.payload
-    }
+            #[inline]
+            fn payload(&self) -> &Payload {
+                &self.payload
+            }
+        }
+
+        #[cfg(feature = "std")]
+        impl SignableTransaction for $ty {
+            fn sign(
+                mut self,
+                key_pair: iroha_crypto::KeyPair,
+            ) -> Result<SignedTransaction, iroha_crypto::Error> {
+                let signature = SignatureOf::new(key_pair, &self.payload)?;
+                self.signatures.insert(signature);
+
+                Ok(SignedTransaction {
+                    payload: self.payload,
+                    signatures: self.signatures,
+                })
```

Comment:
I think, you can use new signature collection instead of using the one from `self`. Then this will not confuse peaple, then they read that *Transaction without signatures*, but struct actually stores `signatures`

Reply:
Oh, I see, you need to store `signatures` in `Transaction` to be able to use this macro with `Transaction` and `SignedTransaction` at the same time....

Reply:
Well, in my opinion it's better to write this impl by hand and not to store meaningless `signatures` in `Transaction`.

If you are not agree, we can wait for other opinions

Reply:
Or you can upgrade your macro to accept default `signatures` state.
Some examples as I see it:

```rust
impl_tx! {
    Transaction<Singatures=BTreeSet::new()>,
    SignedTransaction<Signatures=self.signatures>
}

impl_tx! {
    Transaction where Singatures: BTreeSet::new(),
    SignedTransaction where Signatures: self.signatures
}

impl_tx! {
    Transaction with BTreeSet::new() as signatures,
    SignedTransaction with self.signatures as signatures
}
```

Reply:
i think manual implementation would be enough in this case.